iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/utils/CommandDetector.java

```diff
@@ -78,44 +82,149 @@ public List<Command> search(List<ArgumentResolution> resolvedArguments) {
    */
   // Cognitive Complexity of methods should not be too high
   @SuppressWarnings("java:S3776")
+  // TODO: Rework commentary
   private List<ArgumentResolution> fullMatch(Deque<ArgumentResolution> argumentStack) {
-    List<ArgumentResolution> commandArguments = new ArrayList<>();
-    Deque<CommandPredicate> predicateStack = new LinkedList<>(predicates);
-    while (!predicateStack.isEmpty()) {
-      CommandPredicate currentPredicate = predicateStack.pollFirst();
-      ArgumentResolution resolution = argumentStack.pollFirst();
+    CommandDetectionStatus detectionStatus = new CommandDetectionStatus(argumentStack, new LinkedList<>(predicates), new ArrayList<>());
+    while (!detectionStatus.predicateStack.isEmpty()) {
+
+      detectionStatus.setCurrentPredicate(detectionStatus.predicateStack.pollFirst());
+      // resolution is removed from stack during match-methods, here it is peeked to see if it is null or UNRESOLVED
+      ArgumentResolution resolution = detectionStatus.argumentStack.peekFirst();
 
       // Stop argument detection when argument list is empty
       if (resolution == null) {
-        return remainingPredictsAreOptional(currentPredicate, predicateStack) ? commandArguments : Collections.emptyList();
+        return remainingPredictsAreOptional(detectionStatus.getCurrentPredicate(), detectionStatus.predicateStack) ? detectionStatus.commandArguments : Collections.emptyList();
       }
 
       // Stop argument detection when argument is unresolved to start new command detection
       if (resolution.isUnresolved()) {
+        argumentStack.pop();
         return Collections.emptyList();
       }
 
-      // Test argument resolution with predicate
-      if (currentPredicate.predicate.test(resolution.value())) {
-        // Skip argument and start new command detection
-        if (currentPredicate.is(NO_MATCH)) {
-          return Collections.emptyList();
-        }
-        // Re-add predicate to stack to be reevaluated on the next argument
-        if (currentPredicate.is(ZERO_OR_MORE)) {
-          predicateStack.addFirst(currentPredicate);
-        }
-        // Add matched argument to command
-        commandArguments.add(resolution);
-      } else if (currentPredicate.is(OPTIONAL, ZERO_OR_MORE, NO_MATCH)) {
-        // Re-add argument to be evaluated by the next predicate
-        argumentStack.addFirst(resolution);
+      if (detectionStatus.getCurrentPredicate() instanceof SingularPredicate) {
```

Comment:
As you call the methods 7 times inside the loop maybe store its value in a variable as it is not expected to return different values withing a single loop iteration.